File: uapc/archive/2022/problems/clumsycardinals/submissions/accepted/clumsycardinals-zf.py
#!/usr/bin/python3

def find(u):
    if uf[u] != u: uf[u] = find(uf[u])
    return uf[u];

def union(u, v):
    u, v = find(u), find(v)
    if u != v:
        uf[u] = v
        return 1
    else:
        return 0

comp = n = int(input())

uf, ldiag, rdiag = [i for i in range(n)], {}, {}

vals = list(map(int, input().split()))

for i in range(n):
    x, y = vals[2*i], vals[2*i+1]

    ld, rd = x-y, x+y

    tup = (x, y, i)

    for j in range(-1, 2):
        if ld+j in ldiag: comp -= union(i, ldiag[ld+j][2])
        if rd+j in rdiag: comp -= union(i, rdiag[rd+j][2])
    ldiag[ld] = rdiag[rd] = tup

print(comp)

File: uapc/archive/2022/problems/clumsycardinals/submissions/accepted/clumsycardinals-zf.cpp
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

vector<int> uf;

struct point {
  int x, y, i;
};

int find(int u) {
   return u == uf[u] ? u : uf[u] = find(uf[u]);
}

int merge(int u, int v) {
  u = find(u);
  v = find(v);
  uf[u] = v;
  return u != v;
}

int main() {
  int n;
  cin >> n;
  int comp = n;

  unordered_map<int, point> ldiag, rdiag;

  for (int i = 0; i < n; ++i) {
    uf.push_back(i);

    point p;
    cin >> p.x >> p.y;
    p.i = i;

    int ld = p.x-p.y, rd = p.x+p.y;

    if (ldiag.find(ld) != ldiag.end()) {
      comp -= merge(i, ldiag[ld].i);
    }
    else {
      ldiag[ld] = p;
      for (int j = -1; j <= 1; j += 2)
        if (ldiag.find(ld+j) != ldiag.end())
          comp -= merge(i, ldiag[ld+j].i);
    }

    if (rdiag.find(rd) != rdiag.end()) {
      comp -= merge(i, rdiag[rd].i);
    }
    else {
      rdiag[rd] = p;
      for (int j = -1; j <= 1; j += 2)
        if (rdiag.find(rd+j) != rdiag.end())
          comp -= merge(i, rdiag[rd+j].i);
    }
  }

  cout << comp << endl;

  return 0;
}
